cas-server-support-token/src/main/java/org/jasig/cas/authentication/handler/support/TokenAuthenticationHandler.java

```diff
@@ -0,0 +1,88 @@
+package org.jasig.cas.authentication.handler.support;
+
+import org.apache.commons.lang3.StringUtils;
+import org.jasig.cas.TokenConstants;
+import org.jasig.cas.authentication.Credential;
+import org.jasig.cas.authentication.HandlerResult;
+import org.jasig.cas.authentication.PreventedException;
+import org.jasig.cas.authentication.handler.PrincipalNameTransformer;
+import org.jasig.cas.integration.pac4j.authentication.handler.support.TokenWrapperAuthenticationHandler;
+import org.jasig.cas.services.RegisteredService;
+import org.jasig.cas.services.RegisteredServiceProperty;
+import org.jasig.cas.services.UnauthorizedServiceException;
+import org.pac4j.jwt.credentials.authenticator.JwtAuthenticator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import java.security.GeneralSecurityException;
+
+/**
+ * This is {@link TokenAuthenticationHandler} that authenticates instances of {@link TokenCredential}.
+ * There is no need for a separate {@link org.jasig.cas.authentication.principal.PrincipalResolver} component
+ * as this handler will auto-populate the principal attributes itself.
+ *
+ * @author Misagh Moayyed
+ * @since 4.2.0
+ */
+@Component("tokenAuthenticationHandler")
+public class TokenAuthenticationHandler extends TokenWrapperAuthenticationHandler {
+
+    @Override
+    protected final HandlerResult doAuthentication(final Credential credential) throws GeneralSecurityException, PreventedException {
+        final TokenCredential tokenCredential = (TokenCredential) credential;
+        logger.debug("Locating token secret for service [{}]", tokenCredential.getService());
+
+        final RegisteredService service = this.servicesManager.findServiceBy(tokenCredential.getService());
+        final String tokenSecret = getRegisteredServiceJwtSecret(service);
+
+        if (StringUtils.isNotBlank(tokenSecret)) {
+            final JwtAuthenticator tokenAuthenticator = new JwtAuthenticator(tokenSecret);
+            setAuthenticator(tokenAuthenticator);
```

Comment:
Multi-threading issue I think: you get a specific jwtToken for a registered service, but you only have one shared authenticator for all tokens and services. 

I guess the pac4j wrapper classes hierarchy must be changed to pass the appropriate authenticator for login.